### PR TITLE
Onboarding Playwright (JavaScript and TypeScript) to AzureTestPlanV0 Task

### PR DESCRIPTION
**Pull request by anshika-gupta21**
### **Task Name**: AzureTestPlanV0

---

### **Description**

Currently, ATP supports four frameworks: JavaMaven, JavaGradle, PyTest, Jest. This PR introduces the integration of the Playwright framework, supporting both JavaScript and TypeScript, with Azure Test Plans (ATP) Task to enhance automated testing capabilities. Playwright is a powerful open-source automation framework that simplifies end-to-end testing across different browsers.
### **Key Features & Implementation Details:**

1. **ATP Integration**:

- The pipeline is configured to connect Playwright tests with ATP, enabling automated execution of test plans and test cases.

2. **Test Execution & Reporting**:

- The pipeline executes Playwright tests and publishes results in the tests tab.
- Execution status and history are reflected in the Execute tab of the test plan.
- The Runs tab displays detailed test reports, including the attached test-results.xml file.
- Detailed reporting includes pass/fail status, execution time, and browser-specific results.

3. **Cross-Package Test Handling and Telemetry**:

- Test cases located in different packages under the tests folder are automatically discovered and executed.
- The Playwright test executor uses grep to collect all test cases and generate a unified JUnit XML report.
- Pipeline logs include telemetry to support traceability and debugging.
- Playwright-specific test modifiers such as .skip(), .fail(), and .only() are handled to reflect expected outcomes.

4. **Multi-Browser and Dual Language Support**:

- Tests are executed across Chromium, Firefox, and WebKit to ensure cross-browser compatibility.
- Both JavaScript and TypeScript test files are supported while using a common executor in the background.

---

### **Demo video**:  https://microsoftapc-my.sharepoint.com/:v:/g/personal/t-anshgupta_microsoft_com/EZiGK7aFm-9FgpELYZyNQxsBFzmUgRUTh5uVDBeys_drtg?e=IMknGX
---
### **Sample Pipeline to run**: https://dev.azure.com/tfspfcusctest/Azure%20Test%20Plan%20Task%20Playground/_build/results?buildId=10749&view=results
---
The language selection options visible in UI: 

ui_ss2

Published results on Test tab:

tests tab

test-results.xml file:

testresultsxml

Associated Automation:

associatedAutomation

Progress Report: 

progress_report

---

### **Checklist**
- [x] Task version was bumped — see [versioning guide](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md)
- [x] Verified the task behaves as expected

### CODE REVIEW

File: Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            tl.debug("Error in setting up env for Playwright Test Framework: " + operationResult.errorMessage);
+        }
+
+        return operationResult;
+    }
+
+    async discoverTests(listOfTestsToBeExecuted: string[], ciData: ciDictionary, listOfTestsToBeRan: string[]): Promise<IOperationResult> {
+        let operationResult: IOperationResult = { returnCode: 0, errorMessage: '' };
+
+        listOfTestsToBeExecuted.forEach(element => {
```
Comment (bhaveshbhati) [inline Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts]: nit: test instead of element

File: Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    async executeTests(testsToBeExecuted: string[], ciData: ciDictionary): Promise<IOperationResult> {
+    let operationResult: IOperationResult = { returnCode: 0, errorMessage: '' };
+    let executionTimer = new SimpleTimer(constants.AUTOMATED_EXECUTION);
+
+    tl.debug('Entering executeTests method');
+    tl.debug(`Tests to be executed: ${JSON.stringify(testsToBeExecuted)}`);
+
+    if (!testsToBeExecuted || testsToBeExecuted.length === 0) {
+        tl.debug('No tests to execute');
+        return { returnCode: 0, errorMessage: 'No tests to execute' };
```
Comment (bhaveshbhati) [inline Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts]: nit: formatting

File: Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        });
+
+        return operationResult;
+    }
+
+    async executeTests(testsToBeExecuted: string[], ciData: ciDictionary): Promise<IOperationResult> {
+    let operationResult: IOperationResult = { returnCode: 0, errorMessage: '' };
+    let executionTimer = new SimpleTimer(constants.AUTOMATED_EXECUTION);
+
+    tl.debug('Entering executeTests method');
```
Comment (bhaveshbhati) [inline Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts]: I think this can be removed.

File: Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        if (grepArg) {
+            this.toolRunner.arg('-g');
+            this.toolRunner.arg(grepArg);
+        }
+
+        const commandPreview = `npx cross-env PLAYWRIGHT_JUNIT_OUTPUT_NAME=${junitOutput} playwright test --reporter=junit -g "${grepArg}"`;
+        tl.debug(`Executing Playwright test command: ${commandPreview}`);
+
+        operationResult.returnCode = await this.toolRunner.execAsync();
+
```
Comment (bhaveshbhati) [inline Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts]: nit: remove extra line.

File: Tasks/AzureTestPlanV0/Automated Flow/automatedFlow.ts [hunk 5]
```diff
@@ -158,6 +158,10 @@ function getTestExecutor(testLanguage: string): ITestExecutor {
             return new PythonTestExecutor();
         case 'javascriptjest':
             return new JestTestExecutor();
+        case 'javascriptplaywright': 
+            return new PlaywrightTestExecutor();
+        case 'typescriptplaywright':
+            return new PlaywrightTestExecutor();
```
Comment (adityashahms) [inline Tasks/AzureTestPlanV0/Automated Flow/automatedFlow.ts:162]: If understand correctly, both are doing the identical work. Can we use a single option instead of providing two?

File: Tasks/AzureTestPlanV0/Common/utils.ts [hunk 6]
```diff
@@ -73,6 +73,17 @@ export function separateJestTestName(inputString) {
     }
 }
 
+export function separatePlaywrightTestName(inputString) {
```
Comment (adityashahms) [inline Tasks/AzureTestPlanV0/Common/utils.ts:80]: Add comment before the function showcasing input and output of function

File: Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    if (!testsToBeExecuted || testsToBeExecuted.length === 0) {
+        tl.debug('No tests to execute');
+        return { returnCode: 0, errorMessage: 'No tests to execute' };
+    }
+
+    executionTimer.start();
+
+    let grepArg = '';
+    try {
+        const junitOutput = 'test-results/test-results.xml';
```
Comment (adityashahms) [inline Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts]: Pull out all the code used for setup before the start() function invocation on line 71. In the execution timer, lets track only the time used for executing test not for building the cmdlet.

File: Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        this.toolRunner = tl.tool(this.toolRunnerPath);
+
+        // Building the command: cross-env PLAYWRIGHT_JUNIT_OUTPUT_NAME=... playwright test --reporter=junit -g ...
+        this.toolRunner.arg('cross-env');
+        this.toolRunner.arg(`PLAYWRIGHT_JUNIT_OUTPUT_NAME=${junitOutput}`);
+        this.toolRunner.arg('playwright');
+        this.toolRunner.arg('test');
+        this.toolRunner.arg('--reporter=junit');
+
+        if (grepArg) {
```
Comment (adityashahms) [inline Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts]: Is check required? Since we would have the grep argument each time? In line 66, we already are checking if we have list of valid test cases or not

File: Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    try {
+        const junitOutput = 'test-results/test-results.xml';
+        tl.setVariable('PLAYWRIGHT_JUNIT_OUTPUT_NAME', junitOutput);
+        const resultsDir = path.join(process.cwd(), 'test-results');
+        if (!fs.existsSync(resultsDir)) {
+            fs.mkdirSync(resultsDir);
+        }
+
+        // Playwright test name selection usually uses 'grep'
+        if (testsToBeExecuted && testsToBeExecuted.length > 0) {
```
Comment (adityashahms) [inline Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts]: If condition is redundant since we are checking before hand that testsToBeExecuted is nonnull and nonempty

File: Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+            failureType: 'Execution Error'
+        };
+    }
+
+    executionTimer.stop(ciData);
+    ciData['grepArgument'] = grepArg;
+    ciData['executionStatus'] = operationResult.returnCode === 0 ? 'Success' : 'Failure';
+
+    // Log telemetry data
+    tl.debug(`Telemetry Data Before Logging: ${JSON.stringify(ciData)}`);
```
Comment (bhaveshbhati) [inline Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts]: nit: Should we log this ?

File: Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    * Setup the test executor
+    */
+    async setup(): Promise<IOperationResult> {
+        let operationResult: IOperationResult = { returnCode: 0, errorMessage: '' };
+
+        try {
+            
+            this.toolRunnerPath = tl.which(this.testRunnerCLI, true);
+            this.toolRunner = tl.tool(this.toolRunnerPath);
+            this.toolRunner.line('install');
```
Comment (KathanS) [inline Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts:27]: Should we also check if cross-env is present using tl.which('cross-env', false) and install if not present? since, we will be using cross-env later

File: Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        const junitOutput = 'test-results/test-results.xml';
+        tl.setVariable('PLAYWRIGHT_JUNIT_OUTPUT_NAME', junitOutput);
+        const resultsDir = path.join(process.cwd(), 'test-results');
+        if (!fs.existsSync(resultsDir)) {
+            fs.mkdirSync(resultsDir);
+        }
+
+        // Playwright test name selection usually uses 'grep'
+        if (testsToBeExecuted && testsToBeExecuted.length > 0) {
+            const grepPattern = testsToBeExecuted.map(t => `${t}`).join('|');
```
Comment (KathanS) [inline Tasks/AzureTestPlanV0/Automated Flow/TestExecutors/PlaywrightTestExecutor.ts]: small suggestion on edge case: should we add escape for regex, as in add \ before special characters? for example, if test name is "[test]" it should be passed as "\\[test\\]" in grepArg to avoid "[", ']" considered as special characters used in regex matching

**Comment by KathanS** [comment]
@anshika-gupta21 can you please merge the latest changes from master into this branch? we can run pending checks after updating the branch

**Comment by anshika-gupta21** [comment]
> @anshika-gupta21 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="Microsoft"

**Review by adityashahms — APPROVED** [review]
LGTM
